### ISSUE TITLE
Bootstrap FATAL: Does not have an entry for the last committed index

### ISSUE THREAD

**Issue by ttyusupov** [yugabyte/yugabyte-db#2565]
I've faced this issue often while reproducing #2563. After sometime (seconds-minutes) is passed after node hit memory limits due to incoming RPC memory consumption it is often getting in the following error:
```
F1009 22:12:43.026485 26269 tablet_bootstrap.cc:969] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea: Does not have an entry for the last committed index: { term: 16 index: 12915 }, entries: [{12654, { entry: 0x000000001f65be00 -> type: REPLICATE replicate { id { term: 16 index: 12654 } hybrid_time: 6433418851954266112 op_type: WRITE_OP write_request { tablet_id: "783e193b19654cbb8706c1983fe8c282" write_batch { write_pairs { key: "G\270\336Smetric-12188\000\000
```
Restarting `yb-tserver` is not helping - it quickly gets into this FATAL again. Only removing data for the broken tablets on this node helps.

Steps for reproducing: https://github.com/yugabyte/yugabyte-db/issues/2566#issuecomment-549510008

**Comment by ttyusupov** [comment]
[yb-tserver.yb-admin-timur-test2-n1.yugabyte.log.INFO.20191009-214520.23800.gz](https://github.com/yugabyte/yugabyte-db/files/3713842/yb-tserver.yb-admin-timur-test2-n1.yugabyte.log.INFO.20191009-214520.23800.gz)
[yb-tserver.yb-admin-timur-test2-n1.yugabyte.log.INFO.20191009-220440.25195.gz](https://github.com/yugabyte/yugabyte-db/files/3713843/yb-tserver.yb-admin-timur-test2-n1.yugabyte.log.INFO.20191009-220440.25195.gz)
[yb-tserver.yb-admin-timur-test2-n1.yugabyte.log.INFO.20191009-221107.26256.gz](https://github.com/yugabyte/yugabyte-db/files/3713844/yb-tserver.yb-admin-timur-test2-n1.yugabyte.log.INFO.20191009-221107.26256.gz)

**Comment by bmatican** [comment]
Still not able to get to the bottom of the FATAL, however, found some potential perf issues that might be nice to tackle!

1) Doing enough applys to rocksdb that we stop writes due to 2 memtables
```
7145:I1009 22:12:24.141655 26269 tablet_bootstrap_if.cc:75] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea: Bootstrap replayed 1/3 log segments. Stats: ops{read=1196 overwritten=0} inserts{seen=0 ignored=0} mutations{seen=0 ignored=0}. Pending: 1 replicates
7397:I1009 22:12:28.712260 26269 db_impl.cc:5178] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea [R]: [default] New memtable created with log file: #171
7398:W1009 22:12:28.722470 26269 column_family.cc:674] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea [R]: [default] Increasing compaction threads because we have 12 level-0 files·
8183:I1009 22:12:30.174646 26345 flush_job.cc:249] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea [R]: [default] [JOB 3] Flushing memtable with next log file: 171
8184:I1009 22:12:30.174671 26345 event_logger.cc:67] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea [R]: EVENT_LOG_v1 {"time_micros": 1570659150174663, "job": 3, "event": "flush_started", "num_memtables": 1, "num_entries": 225000, "num_deletes": 0, "memory_usage": 16805592}
8185:I1009 22:12:30.174685 26345 flush_job.cc:277] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea [R]: [default] [JOB 3] Level-0 flush table #174: started
8197:I1009 22:12:30.425673 26269 db_impl.cc:5178] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea [R]: [default] New memtable created with log file: #171
8198:W1009 22:12:30.436471 26269 column_family.cc:598] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea [R]: [default] Stopping writes because we have 2 immutable memtables (waiting for flush), max_write_buffer_number is set to 2
```

2) I think we might be applying to rocksdb ALL of the entries from the WAL files

Log entry:
> 3869:I1009 22:11:48.003137 26269 tablet_bootstrap.cc:841] T 783e193b19654cbb8706c1983fe8c282 P 2f382859ba954bdeb0e8cb1730ab9dea: Max persistent index in RocksDB's SSTables before bootstrap: term: 16 index: 11818/term: -1 index: -1

Code that is supposed to skip applying:
```
  auto min_index = std::min(
      state->regular_stored_op_id.index(), state->intents_stored_op_id.index());
  if (op_id.index() <= min_index) {
    // Do not update the bootstrap in-memory state for log records that have already been applied to
    // RocksDB, or were overwritten by a later entry with a higher term that has already been
    // applied to both regular and provisional record RocksDB.
    replicate_entry_ptr->reset();
    return Status::OK();
  }
```

But for non-txn tables, intents will be -1, which I assume compares as smaller than valid op ids. This would imply we never skip WAL entries for non-txn tables! cc @ttyusupov @spolitov 

In particular if 2) makes sense, I can spin that off as a separate investigation.

**Comment by bmatican** [comment]
wrt point 2) above, @spolitov confirmed earlier that there's a separate check later:
```
Status TabletBootstrap::HandleEntryPair(
...
  if (replicate->id().index() > flushed_index) {
    const auto status = HandleOperation(op_type, replicate);
```
This takes care of ignoring smaller op ids.

For point 1) I filed #2619 to track separately.

**Comment by bmatican** [comment]
Apparently we're hitting these in our tests as well! Should dig further into this cc @rahuldesirazu 

```
cat fatals_600 | xargs grep -l 'Does not have an entry for the last committed index'
/var/lib/jenkins/jobs/github-yugabyte-db-centos-master-clang-tsan/builds/611/archive/build/tsan-clang-dynamic-ninja/yb-test-logs/tests-client__ql-transaction-test/QLTransactionTest_WriteConflictsWithRestarts.log
/var/lib/jenkins/jobs/github-yugabyte-db-centos-master-clang-tsan/builds/617/archive/build/tsan-clang-dynamic-ninja/yb-test-logs/tests-client__snapshot-txn-test/SnapshotTxnTest_MultiWriteWithRestart.log
/var/lib/jenkins/jobs/github-yugabyte-db-centos-master-clang-tsan/builds/654/archive/build/tsan-clang-dynamic-ninja/yb-test-logs/tests-client__snapshot-txn-test/SnapshotTxnTest_MultiWriteWithRestart.log
/var/lib/jenkins/jobs/github-yugabyte-db-centos-master-clang-tsan/builds/665/archive/build/tsan-clang-dynamic-ninja/yb-test-logs/tests-client__snapshot-txn-test/SnapshotTxnTest_MultiWriteWithRestart.log
/var/lib/jenkins/jobs/github-yugabyte-db-centos-master-clang-tsan/builds/672/archive/build/tsan-clang-dynamic-ninja/yb-test-logs/tests-client__snapshot-txn-test/SnapshotTxnTest_MultiWriteWithRestart.log
/var/lib/jenkins/jobs/github-yugabyte-db-centos-master-clang-tsan/builds/680/archive/build/tsan-clang-dynamic-ninja/yb-test-logs/tests-client__snapshot-txn-test/SnapshotTxnTest_MultiWriteWithRestart.log
```

**Comment by bmatican** [comment]
Likely fixed by 89563a52cbdb38f4db989082d6deb0cbedd3bdbd now
